### ISSUE TITLE
Add flag for load() and preload() that lets godot also load any resources within the scenes. 

### ISSUE THREAD

**Issue by TheConceptBoy** [godotengine/godot-proposals#10435]
### Describe the project you are working on

A game where cutscenes are created as TSCN scenes and need to be seamlessly played back when needed.

This is when you want to seamlessly cue up the next scene or a particularly heavy scene that needs to be played back immediately.

### Describe the problem or limitation you are having in your project

When a scene is loaded using load() or preload(), godot does not actually load any of the resources that make up that scene until it is instanced and added as a child. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

An extra flag in the load() and preload() that forces godot to also buffer any resources found within the scene so that when the scene is instantiated, the game does not stutter as it's rushing to load all the resources used in that scene.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

load("res://cutscenes/scene_1.tscn", true) 

first supplied parameter is the scene, second parameter determines if godot should scan the scene for any resources (sprites, sounds, etc) and preload those as well.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No. load and preload are core functions.

### Is there a reason why this should be core and not an add-on in the asset library?

load and preload are core functions.

**Comment by AthertonRoamer** [comment]
Since resources are cached when loaded by the engine, you could manually load all the resources in the scene when you load the scene. Perhaps in a static function owned by the cut scene?

**Comment by AthertonRoamer** [comment]
For the workaround I mentioned above to work, you would have to make sure there is always a reference to the resource or it would be freed since it extends RefCounted. One way to do this would be to use the RefCounted reference() method to increase the resource's reference counter, but this could lead to memory leaks if you didn't unreference later. Something similar would need to be done internally for the feature you are suggesting.

**Comment by ricmzn** [comment]
> When a scene is loaded using load() or preload(), godot does not actually load any of the resources that make up that scene until it is instanced and added as a child.

This is false. Godot already loads all resource dependencies recursively and immediately, though what constitutes a dependency could be unclear, it generally includes:
- Scene instances added in the editor
- `@export` vars filled in inspector
- Built-in properties filled in inspector
- Script preload() calls (written anywhere in a script that is being loaded, even if in a function that never runs)
- And of course, dependencies of dependencies

You can verify this, for example, by adding custom resources to any of your scene's dependencies, that print something on _init() (called when the resource is loaded and instantiated). RAM and VRAM usage will also go up very clearly when loading a scene that depends on many other resources, especially textures and audio files.

But load() calls are dynamic, and as such, don't begin loading until the load() call actually happens during script execution.

If you are seeing stutters when adding instances of a scene to the tree, and you are sure there are no dynamic load() calls in it, the issue is likely something else, like collision BVH updates or shader compilations, both of which currently run on the main thread with no way to avoid them (shader compilations are generally reduced by Ubershaders on Mobile and Forward+ renderers, but they're not foolproof, especially if materials are modified on the fly).

Heavy _enter_tree() or _ready() functions in the script can also cause add_child() stutters, but at least you can measure and optimize those functions if needed.

**Comment by Calinou** [comment]
> (shader compilations are generally reduced by Ubershaders on Mobile and Forward+ renderers, but they're not foolproof, especially if materials are modified on the fly).

Check the [documentation](https://docs.godotengine.org/en/stable/tutorials/performance/pipeline_compilations.html) for details on how to best make use of ubershaders and the shader baker on export.
